src/dataflow/stream/read_stream.rs

```diff
@@ -187,12 +202,58 @@ impl<D: Data> ReadStream<D> {
         // Poll for the next message
         match self.recv_endpoint.as_mut() {
             Some(endpoint) => match endpoint.read().await {
-                Ok(msg) => Ok(msg),
+                Ok(msg) => {
+                    self.update_condition_context(&msg);
+                    Ok(msg)
+                }
                 // TODO: better error handling.
                 _ => Err(ReadError::Disconnected),
             },
             None => Err(ReadError::Disconnected),
         }
+
+        // TODO: Close the stream? 
+    }
+
+    /// Maintains the statistics for the ConditionContext.
+    fn update_condition_context(&mut self, msg: &Message<D>) {
+        slog::debug!(
+            crate::TERMINAL_LOGGER,
+            "Updating statistics for the message: {:?}",
+            msg
+        );
+        match msg {
+            Message::TimestampedData(td) => {
+                // Increment the message count.
+                self.condition_context
+                    .increment_msg_count(self.id(), td.timestamp.clone());
+            }
+            Message::Watermark(t) => {
+                // Notify the watermark arrival.
+                self.condition_context
+                    .notify_watermark_arrival(self.id(), t.clone());
+            }
+        }
+    }
+
+    /// Clears the condition context state.
+    pub fn clear_condition_state(&mut self, stream_id: StreamId, timestamp: Timestamp) {
+        self.condition_context.clear_state(stream_id, timestamp);
+    }
+
+    /// Evaluate a condition function on the statistics of the ReadStream.
+    pub fn evaluate_condition(&self, condition_fn: &CondFn) -> bool {
```

Comment:
I'm not sure this is the right place to do this in terms of abstractions. Can we move this to the `OperatorExecutor` and call `get_condition_context` to evaluate the condition from there?

Reply:
I agree, but there's a mismatch between the API on the `ReadStream` and `WriteStream` since we cannot get a `&ConditionContext` from the write stream.

I'm gonna make the change for now since these functions will remain `pub(crate)`, but once we have the rest of the deadlines implemented, we should probably revisit the correct API for this.